src/consensus/pbft/libbyz/Replica.cpp

```diff
@@ -269,64 +288,60 @@ bool Replica::compare_execution_results(
   return true;
 }
 
-bool Replica::apply_ledger_data(const std::vector<uint8_t>& data)
+void Replica::playback_request(const pbft::Request& request)
 {
-  PBFT_ASSERT(ledger_replay, "ledger_replay should be initialized");
-
-  if (data.empty())
+  auto req =
+    create_message<Request>(request.raw.data(), request.raw.size()).release();
+  if (!(req->size() > Request::big_req_thresh && brt.add_request(req)))
   {
-    LOG_FAIL << "Received empty entries" << std::endl;
-    return false;
+    rqueue.append(req);
   }
+}
 
-  auto executable_pps = ledger_replay->process_data(
-    data, rqueue, brt, *ledger_writer.get(), last_executed);
+void Replica::playback_pre_prepare(const pbft::PrePrepare& pre_prepare)
+{
+  auto executable_pp = create_message<Pre_prepare>(
+    pre_prepare.contents.data(), pre_prepare.contents.size());
+  auto seqno = executable_pp->seqno();
 
-  for (auto& executable_pp : executable_pps)
+  ByzInfo info;
+  if (execute_tentative(executable_pp.get(), info))
   {
-    auto seqno = executable_pp->seqno();
-
-    ByzInfo info;
-    if (execute_tentative(executable_pp.get(), info))
+    auto batch_info = compare_execution_results(info, executable_pp.get());
```

Comment:
Minor: this can be `if(compare_execution_results(info, executable_pp.get()) { return; }`.